x/src/cargo.rs

```diff
@@ -86,39 +107,65 @@ impl Cargo {
         self
     }
 
+    // Passes extra environment variables to be passed to the executable.
+    pub fn env_additions<I, S>(&mut self, envs: I) -> &mut Self
+    where
+        I: IntoIterator<Item = (S, S)>,
+        S: AsRef<OsStr>,
+        //        S: std::convert::AsRef<std::ffi::OsStr>,
+    {
+        self.inner.envs(envs);
+        self
+    }
+
     pub fn run(&mut self) -> Result<()> {
         self.inner.stdout(Stdio::inherit()).stderr(Stdio::inherit());
         self.do_run(true).map(|_| ())
     }
 
+    // This execution will in general not produce std out file output, but will
+    // instead return a byte Vec containing the execution's stdout messages.
     pub fn run_with_output(&mut self) -> Result<Vec<u8>> {
         self.inner.stderr(Stdio::inherit());
         // Since system out hijacked don't log for this command
```

Comment:
`// Since stdout is captured, don't log this command.`

Reply:
But I don't really understand the logic here -- why does stdout being captured mean that the command isn't logged? Could you elaborate? Thanks :D

Reply:
Internally x has to run some commands to figure out where it's being executed, and find the workspace location.   Those would be noise to the average user -- they are the run commands called with the Vec<u8> output - to facilitate internal processing.    I'm logging out the commands that a user might want to run on their own - with relevant timing and passed through environment variables.

Reply:
Hmm, that makes sense but it actually feels orthogonal to whether stdout is captured to me.

Reply:
Yeah, just didn't want to try to refactor everything in one go.   This is a result of some of the original structure.   Been commented on and asked about a few times.   I should probably just create a commit renaming the methods, and try to discover if standard out is redirected so I don't need to pass in a parameter to determine if logging should occur.    Not happy with the structure.  Not trusting myself to refactor it to what I want yet.   The techniques I really want would involve reflection/generic method wrappers.

Reply:
This sort of goes back to what I initially said, if you're going to log a command, log every single command. If you think it'll be too much noise for the average user then we should not be logging those things by default (at a level where the avg user would see them) but rather allow people to turn on an appropriate `-v` or RUST_LOG in order to see those messages.

Reply:
I've written up here where I want to take the design.  https://github.com/libra/libra/pull/3261#discussion_r403799067. Was trying to move that discussion out of this pr as it will change with needs.